File: 2021/fsharp/Aoc2021/Day03.fs
module Day03

open Utils

type Bits =
    | ZeroesWin
    | OnesWin

module Bits =
    let count (lines: string list) (index: int) =
        lines
        |> List.sumBy (fun line -> if line.[index] = '1' then 1 else -1)
        |> System.Math.Sign
        |> function
            | -1 -> ZeroesWin
            | _ -> OnesWin

    let toChar =
        function
        | ZeroesWin -> '0'
        | OnesWin -> '1'

    let not =
        function
        | ZeroesWin -> OnesWin
        | OnesWin -> ZeroesWin

let mostCharAt (lines: string list) (index: int) =
    index |> Bits.count lines |> Bits.toChar

let leastCharAt (lines: string list) (index: int) =
    index |> Bits.count lines |> Bits.not |> Bits.toChar

let bitstringToInt bits = System.Convert.ToInt32(bits, 2)

let joinChars (chars: char list) = System.String.Join("", chars)

type PowerConsumtionRate = { Gamma: int; Epsilon: int }

module PowerConsumtionRate =
    let result (rate: PowerConsumtionRate) = rate.Gamma * rate.Epsilon

    let ofReport (report: string list) =
        let numChars = report.Head.Length

        let gamma =
            [ 0 .. numChars - 1 ]
            |> List.map (mostCharAt report)
            |> joinChars
            |> bitstringToInt

        let epsilon = (~~~gamma) &&& ((1 <<< numChars) - 1)

        { Gamma = gamma; Epsilon = epsilon }

type LifeSupportRate = {
    OxygenGenerator: int
    CO2Scrubber: int
}

module LifeSupportRate =
    let result (rate: LifeSupportRate) = rate.OxygenGenerator * rate.CO2Scrubber

    let private _filter func lines =
        let rec _filter_rec func index (lines: string list) =
            match lines |> List.tryExactlyOne with
            | Some line -> line
            | None ->
                let useChar = func lines index

                lines
                |> List.filter (fun line -> line.[index] = useChar)
                |> _filter_rec func (index + 1)

        _filter_rec func 0 lines

    let ofReport (report: string list) = {
        OxygenGenerator = report |> _filter mostCharAt |> bitstringToInt
        CO2Scrubber = report |> _filter leastCharAt |> bitstringToInt
    }


let report = "03" |> readInputLines |> Seq.toList

let job1 () =
    report
    |> PowerConsumtionRate.ofReport
    |> PowerConsumtionRate.result
    |> string

let job2 () =
    report |> LifeSupportRate.ofReport |> LifeSupportRate.result |> string

File: 2021/csharp/Aoc2021/Day03.cs
using System.Text;

namespace Aoc2021;

public class Day03 : IDay
{
    public override string Day { get; } = nameof(Day03)[3..];

    public override string Result1()
    {
        return PowerConsumtionRate
            .FromReport(InputLines.ToList())
            .Result
            .ToString(CultureInfo.InvariantCulture);
    }

    public override string Result2()
    {
        return LifeSupportRate
            .FromReport(InputLines.ToList())
            .Result
            .ToString(CultureInfo.InvariantCulture);
    }

    public record PowerConsumtionRate(int Gamma, int Epsilon)
    {
        public int Result => Gamma * Epsilon;

        public static PowerConsumtionRate FromReport(List<string> lines)
        {
            int numChars = lines[0].Length;

            StringBuilder gammaBuilder = new();
            for (int c = 0; c != numChars; c++)
            {
                _ = gammaBuilder.Append(MostCharAt(c, lines));
            }

            int gamma = BitstringToInt(gammaBuilder.ToString());
            int epsilon = (~gamma) & ((1 << numChars) - 1);
            return new(gamma, epsilon);
        }
    }

    public record LifeSupportRate(int OxygenGenerator, int CO2Scrubber)
    {
        public int Result => OxygenGenerator * CO2Scrubber;

        public static LifeSupportRate FromReport(List<string> lines)
        {
            string oxygenGeneratorString = filter(lines, MostCharAt);
            string co2ScrubberString = filter(lines, LeastCharAt);

            string filter(List<string> lines, Func<int, List<string>, char> fun)
            {
                for (int c = 0; lines.Count > 1; c++)
                {
                    char useChar = fun(c, lines);
                    lines = lines.Where(line => line[c] == useChar).ToList();
                }

                return lines[0];
            }

            int oxygenGenerator = BitstringToInt(oxygenGeneratorString);
            int co2Scrubber = BitstringToInt(co2ScrubberString);
            return new(oxygenGenerator, co2Scrubber);
        }
    }

    private static char MostCharAt(int index, List<string> lines)
    {
        return Math.Sign(lines.Sum(line => line[index] == '1' ? 1 : -1)) switch
        {
            1 => '1',
            -1 => '0',
            _ => '1',
        };
    }

    private static char LeastCharAt(int index, List<string> lines)
    {
        return Math.Sign(lines.Sum(line => line[index] == '1' ? 1 : -1)) switch
        {
            1 => '0',
            -1 => '1',
            _ => '0',
        };
    }

    private static int BitstringToInt(string bits)
    {
        return Convert.ToInt32(bits, 2);
    }
}
